### ISSUE TITLE
Not webpack/browserify friendly

### ISSUE THREAD

**Issue by Cellule** [ansman/validate.js#12]
I checked the code a bit and realized that the require function is cached, modified and wrapped. Is it really necessary ? Since the actual require is wrapped in a try catch, why not simple use the require function and catch if any error occurs?

see the error below
```
WARNING in ./~/validate.js/validate.js
Critical dependencies:
456:13-20 require function is used in a way, in which dependencies cannot be statically extracted
 @ ./~/validate.js/validate.js 456:13-20
```
You can see more info on good practices here https://github.com/umdjs/umd

**Comment by Cellule** [comment]
I can check into fixing the issue for webpack, unfortunately I have no experience with browserify. I think providing the libraries could be a good solution. I will let you decide on this. I know that for webpack it is not a problem, it will detect and handle correctly if the library is available or not (and warn you).

**Comment by ansman** [comment]
@alexkaraman85 This is a tricky one.
I would rather not have a setup required before you can use the library.

I'm guessing the problem stems from the `require` function only existing in compile time and not runtime, right?

We could perhaps have support for both, that we first look for `moment` in the `validate` object and if that fails we fall back to `require`, would that work for you?

**Comment by mosch** [comment]
@ansman i think webpack will still fail on trying to solve the "reuqire" call in the code.

**Comment by jhnns** [comment]
A more beautiful solution would be to unify moment's and XDate's API into one and use that instead. This would even clean-up your code.

Your users would have to overwrite the `dateParser` to use moment or XDate or whatever they like:

```javascript
validate.dateParser = function (dateString) {
    return moment.utc(dateString);
};
```

By default you could use the basic `Date` object which is also able to parse some date strings:

```javascript
validate.dateParser = function (dateString) {
    return new Date(dateString).getTime();
};
```

Thus you would even not depend on *any* library if your users only need basic parsing capabilities.

And concerning promises: I'd just use the global `Promise` object, which will be in all the JS engines soon. Just write in your README that your library requires a polyfill for the `Promise` object. There are polyfills by all the popular libraries like q or when.

**Comment by ansman** [comment]
Well, my concern is that it would be more work for people using node.js

How about that we hide the fact that we use `require` from browserify or webpack?
Something like this:
```javascript
var r = root["require"];
if (!v.isFunction(r)) {
  return null;
}
// Regular require code
```

This way it would work if you have an AMD loader in your browser and on node but on tools that parse your JS to analyze your deps it would think that we have no dependencies.

**Comment by alexkaraman85** [comment]
@ansman 

If you were to hide the requires of validate.js from node.js bundlers then you would lose the Promises and Datetime functionality.  

One other option is that tryRequire goes away and you only use require("text").  This way you code will work for the client and server uses of require.  The downfall is that the bundle will be unnecessarily heavy if the consumer does not use Asynch Validation and/or Datetime.

**Comment by ansman** [comment]
@alexkaraman85 The reason I want to hide the `require` function is because I only want to run it at runtime, not when the static analyser is trying to figure out the dependency graph.

Why wouldn't it work to mask the fact that we use require?
This is the output from node.js
```
> this["require"]
{ [Function: require]
  resolve: [Function],
  main: undefined,
  extensions:
   { '.js': [Function],
     '.json': [Function],
     '.node': [Function: dlopen] },
  registerExtension: [Function],
  cache: {} }
> var validate = this["require"]("./validate.js");
undefined
> validate.validate();
undefined
```

Browserify exposes the `require` function at runtime so this should work there. If webpack doesn't do that then the people who use it will just have to manually inject it like people have suggested.

**Comment by ansman** [comment]
In fact, I just tried it in Browserify and it works as expected. You'd have to manually include for example moment.js if you wanted that but otherwise it picks it up in runtime.

**Comment by jhnns** [comment]
`this["require"]` is only working in node when using the CLI (because `this` is the `global` object then). In a module context, `this` is `module.exports` which has no `require()` function.

But you can write `module["require"]` and that works in node and webpack (haven't tested browserify though).

**Comment by ansman** [comment]
Hmm, `this["require"]` might only work in the CLI but passing require as a function like we do in validate.js works.

I tried running the following things:
```
$ npm install moment
moment@2.9.0 node_modules/moment

$ cat test.js
var validate = require("./validate");
console.log(validate.validators.datetime.parse("2015-03-14"));

$ node test.js
1426291200000

$ npm uninstall moment
unbuild moment@2.9.0

$ node test.js
validate.js:809
        throw new Error("Neither XDate or moment.js was found");
              ^
Error: Neither XDate or moment.js was found
    at Function.validate.validators.datetime.v.extend.parse (validate.js:809:15)
    at Object.<anonymous> (test.js:2:42)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10)
    at startup (node.js:129:16)
    at node.js:814:3
```

**Comment by jhnns** [comment]
It's just that all these trickeries are not necessary imho. Just provide an API where your users can plug-in their favorite date parsing library :)

**Comment by ansman** [comment]
I already have. You can simply set `validate.Promise` to the constructor of the actual promise an override `validate.validators.datetime.parse` and `validate.validators.datetime.format`.

The other stuff is simply convenience methods which is why I would like to suppress the warnings and close this issue.

**Comment by alexkaraman85** [comment]
@ansman what is your need for this requirement?
*The reason I want to hide the require function is because I only want to run it at runtime, not when the static analyser is trying to figure out the dependency graph.*

IMHO: For now it makes sense to leave the code as is and add to the documentation the way to provide the dependency on a date lib.  At the moment the only way I found to use moment.js with browserify is as mentioned up above.

```
 validate.validators.datetime.parse = function(value, options) {
    return +moment.utc(value);
  };

  validate.validators.datetime.format = function(date, options) {
    format = format || (options.dateOnly ? "YYYY-MM-DD" : "YYYY-MM-DD HH:mm:ss");
    return moment.utc(date).format(format);
  };
```

**Comment by ansman** [comment]
Here is a working example of how to get moment working with validate.js and browserify:
```
$ npm install moment validate.js
validate.js@0.5.0 node_modules/validate.js
moment@2.9.0 node_modules/moment

$ cat main.js
var validate = require("validate.js");
console.log(validate.validators.datetime.parse("2015-03-14"));

$ cat index.html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title></title>
</head>
<body>
  <script src="bundle.js"></script>
</body>
</html>

$ browserify -o bundle.js -r moment main.js
```

Now you can simply open `index.html` in your favourite browser and open the console to confirm that it works as expected, no warnings.

As for webpack it does cause a lot of warnings but then you may simply do as @Cellule does and mark it as `noParse` since webpack doesn't seem to support optional dependencies without throwing a bunch of warnings.

I will rethink the dependencies before I release 1.0 but until then things will remain like this. I just don't want this library to require a huge setup when using it.

**Comment by ansman** [comment]
I started playing around with the idea of having to manually set moment, XDate, RSVP etc and it's not that bad, I admit defeat :)

Firstly it removes the need for `tryRequire` and the complex dependency management.
The code would then look like this:
```javascript
var validate = require("validate.js");
validate.moment = require("moment");
console.log(validate.validators.datetime.parse("2015-03-14"));
```

and when running in a browser we simply use the global `moment` object.

I'll make sure to implement this for the next release.

**Comment by ansman** [comment]
Released in 0.6.0

**Comment by jhnns** [comment]
> I started playing around with the idea of having to manually set moment, XDate, RSVP etc and it's not that bad, I admit defeat

:+1: for changing opinions :grinning:
